# Pandas Assignment #2

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('kc_house_data.csv')

In [84]:
df.shape

(21613, 21)

We will start by taking the houses built from 2000 onwards. We will put it in a seperate dataFrame named df_2000.

In [75]:
df_2000=df.loc[df['yr_built']>=2000].copy()

We will check how many houses we have left:

In [83]:
df_2000.shape

(4759, 21)

So there is 4759 houses left.

We need to find the most expensive house for each year. We will put all the results in a new dataFrame named df_exp.
According to that, we will start by putting all the year values in a separate list:
Note that we should remove the duplicates by applying the set function.

In [62]:
yr_built=list(set(df_2000['yr_built'].values))

Now we will create an empty dataFrame and concat the row of the most expensive house for each year:

In [47]:
df_exp=pd.DataFrame()

In [48]:
for y in yr_built:
    df_exp=pd.concat([df_exp, df_2000.loc[df_2000['yr_built']==y].nlargest(1,'price')], axis=0, ignore_index=True)

In [130]:
list_columns=['id','price','bedrooms','bathrooms','sqft_living','sqft_lot','floors','condition','grade','yr_built','waterfront','view','long','lat'] #list of column to display

In [131]:
df_exp[list_columns]

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0   7397300170  3710000.0         4       3.50         5550     28078     2.0   
1   9208900037  6890000.0         6       7.75         9890     31374     2.0   
2   3225069065  3080000.0         4       5.00         4550     18641     1.0   
3   6065300370  4210000.0         5       6.00         7440     21540     2.0   
4   7738500731  4500000.0         5       5.50         6640     40014     2.0   
5   3625049042  3640000.0         5       6.00         5490     19897     2.0   
6   3025059093  3100000.0         5       5.25         5090     23669     2.0   
7   4389201095  3650000.0         5       3.75         5020      8694     2.0   
8   6447300265  4000000.0         4       5.50         7080     16573     2.0   
9   8907500070  5350000.0         5       5.00         8000     23985     2.0   
10  1176001310  2950000.0         5       4.50         4340      5722     3.0   
11  5316100780  2580000.0         4       3.50         3280      3800     2.0   
12  1732800310  2000000.0         4       3.75         2870      4500     2.0   
13  8964800330  3000000.0         4       3.75         5090     14823     1.0   
14  2525049148  3420000.0         5       5.00         5450     20412     2.0   
15  2424059174  2000000.0         4       3.25         5640     35006     2.0   

    condition  grade  yr_built  waterfront  view     long      lat  
0           4     12      2000           0     2 -122.234  47.6395  
1           3     13      2001           0     4 -122.240  47.6305  
2           3     10      2002           1     4 -122.077  47.6053  
3           3     12      2003           0     0 -122.189  47.5692  
4           3     12      2004           1     4 -122.280  47.7493  
5           3     12      2005           0     0 -122.236  47.6165  
6           3     12      2006           0     0 -122.216  47.6297  
7           3     12      2007           0     1 -122.213  47.6146  
8           3     12      2008           0     0 -122.224  47.6151  
9           3     12      2009           0     4 -122.220  47.6232  
10          3     10      2010           0     4 -122.406  47.6715  
11          3     11      2011           0     0 -122.280  47.6299  
12          3     10      2012           0     3 -122.363  47.6291  
13          3     11      2013           0     0 -122.207  47.6200  
14          3     11      2014           0     0 -122.237  47.6209  
15          3     11      2015           0     2 -122.104  47.5491

We need to find the houses for the most bedroom, bathroom and floors respectively. We will put them in seperate dataFrames named respectively df_maxbd , df_maxbth and df_maxflr

In [64]:
df_maxbd=pd.DataFrame()
df_maxbth=pd.DataFrame()
df_maxflr=pd.DataFrame()

In [65]:
for y in yr_built:
    df_maxbd=pd.concat([df_maxbd, df_2000.loc[df_2000['yr_built']==y].nlargest(1,'bedrooms')], axis=0, ignore_index=True)
    df_maxbth=pd.concat([df_maxbth, df_2000.loc[df_2000['yr_built']==y].nlargest(1,'bathrooms')], axis=0, ignore_index=True)
    df_maxflr=pd.concat([df_maxflr, df_2000.loc[df_2000['yr_built']==y].nlargest(1,'floors')], axis=0, ignore_index=True)

We will display the results below:

In [151]:
df_maxbd[list_columns] # The dataFrame of the houses with most bedrooms for each year starting 2000

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0   4139500200  1380000.0         6       4.50         5740     10312     2.0   
1    424069271   936000.0         7       3.75         5100     21802     2.0   
2   3425059222  1300000.0         6       3.50         6563     32670     2.0   
3   7202331220   635000.0         6       2.50         3880      5700     2.0   
4   1778360150  1240000.0         7       5.50         6630     13782     2.0   
5   7852011070  1140000.0         6       3.75         5960     20197     2.0   
6   3021059304   300000.0         6       3.00         2744      9926     2.0   
7   3185600055   495000.0         6       5.00         3440      4500     2.0   
8    627300145  1150000.0        10       5.25         4590     10920     1.0   
9   2149800278   343000.0         6       5.00         2732      7655     2.0   
10  1853081000   820000.0         5       2.75         2830      6137     2.0   
11  6891100260   830000.0         5       3.50         3700      5400     2.0   
12  5609000311   729999.0         6       4.50         3600      6110     2.0   
13  4006000183   450000.0         7       4.00         3150      7800     2.0   
14  2402100575  1130000.0         6       3.75         3010      4360     2.0   
15  9520900210   614285.0         5       2.75         2730      6401     2.0   

    condition  grade  yr_built  waterfront  view     long      lat  
0           3     11      2000           0     2 -122.110  47.5533  
1           3     10      2001           0     0 -122.040  47.5950  
2           3     10      2002           0     0 -122.157  47.6078  
3           3      7      2003           0     0 -122.038  47.6816  
4           3     10      2004           0     0 -122.118  47.5399  
5           3     10      2005           0     4 -121.869  47.5398  
6           3      7      2006           0     0 -122.216  47.2773  
7           3      8      2007           0     0 -122.219  47.4871  
8           3      9      2008           0     2 -122.113  47.5861  
9           3      7      2009           0     0 -122.211  47.3045  
10          3      9      2010           0     0 -122.058  47.5932  
11          3      9      2011           0     0 -122.117  47.7085  
12          3      9      2012           0     0 -122.291  47.5687  
13          3      8      2013           0     0 -122.279  47.5259  
14          3      9      2014           0     0 -122.333  47.6873  
15          3      8      2015           0     0 -122.160  47.7685

In [89]:
df_maxbth[list_columns] # The dataFrame of the houses with most bathrooms for each year starting 2000

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0   2626069030  1940000.0         4       5.75         7220    223462     2.0   
1   9208900037  6890000.0         6       7.75         9890     31374     2.0   
2   3225069065  3080000.0         4       5.00         4550     18641     1.0   
3   6065300370  4210000.0         5       6.00         7440     21540     2.0   
4   1630700380  1920000.0         5       5.75         7730    230868     2.0   
5   3625049042  3640000.0         5       6.00         5490     19897     2.0   
6    622069006  1500000.0         4       5.50         6550    217374     1.0   
7    424069279  1180000.0         6       6.50         6260     10955     2.0   
8   6447300265  4000000.0         4       5.50         7080     16573     2.0   
9   8907500070  5350000.0         5       5.00         8000     23985     2.0   
10  2524069097  2240000.0         5       6.50         7270    130017     2.0   
11  2619950060   465000.0         5       4.00         3210      7200     2.0   
12    42000130   600000.0         5       4.50         4440      9784     2.0   
13   524059063  1800000.0         5       5.00         4490     10279     2.0   
14  2525049148  3420000.0         5       5.00         5450     20412     2.0   
15  7853440050   771005.0         5       4.50         4000      6713     2.0   

    condition  grade  yr_built  
0           3     12      2000  
1           3     13      2001  
2           3     10      2002  
3           3     12      2003  
4           3     12      2004  
5           3     12      2005  
6           3     11      2006  
7           3     11      2007  
8           3     12      2008  
9           3     12      2009  
10          3     12      2010  
11          3      8      2011  
12          3     10      2012  
13          3     10      2013  
14          3     11      2014  
15          3      9      2015

In [92]:
df_maxflr[list_columns] # The dataFrame of the houses with most floors for each year starting 2000

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0   9528102772   438000.0         2       2.00         1270      1372     3.0   
1   7821200390   450000.0         3       2.00         1290      1213     3.0   
2   3300701615   655000.0         4       2.50         2630      4000     3.0   
3   9528104108   535000.0         3       2.50         1360      1016     3.0   
4   6798100662   312000.0         3       1.50         1255      1374     3.0   
5   1972202010   435000.0         3       3.00         1440      1350     3.5   
6   8673400177   525000.0         3       3.00         1730      1074     3.5   
7   1972200426   525000.0         2       2.75         1310      1268     3.5   
8   1702900664   479000.0         2       2.50         1730      1037     3.5   
9   4358700186   275000.0         3       2.25         1260      1488     3.0   
10  3362400431   518500.0         3       3.50         1590      1102     3.0   
11  2559950110  1230000.0         2       2.50         2470       609     3.0   
12  1112000031   715000.0         3       2.25         1990      4977     3.0   
13  5379804150   598800.0         6       4.00         4470     17877     3.0   
14  9528102996   549000.0         3       1.75         1540      1044     3.0   
15  9385200045   729500.0         3       2.50         1660      1091     3.0   

    condition  grade  yr_built  
0           3      8      2000  
1           3      8      2001  
2           3      8      2002  
3           3      7      2003  
4           3      7      2004  
5           3      8      2005  
6           3      8      2006  
7           3      8      2007  
8           3      8      2008  
9           3      7      2009  
10          3      8      2010  
11          3     11      2011  
12          3      9      2012  
13          3      9      2013  
14          3      8      2014  
15          3      9      2015

Next, we will find all the houses with an area larger that the mean area of all the houses. 

Accordingly we will store the mean value of the sqft_living and sqft_lot in 2 variables li_mean and lo_mean:

In [126]:
li_mean=df_2000['sqft_living'].mean()
lo_mean=df_2000['sqft_lot'].mean()

In [78]:
li_mean

2471.24479932759

In [125]:
lo_mean

12238.94788821181

First we will look for the houses with a sqft_living area larger that li_mean: we will store it into a dataframe called df_li:

In [124]:
df_li=df_2000.loc[df_2000['sqft_living']>li_mean].copy()

In [123]:
df_li

id             date      price  bedrooms  bathrooms  \
5      7237550310  20140512T000000  1230000.0         4       4.50   
29     1873100390  20150302T000000   719000.0         4       2.50   
40     5547700270  20140715T000000   625000.0         4       2.50   
42     7203220400  20140707T000000   861990.0         5       2.75   
45     8035350320  20140718T000000   488000.0         3       2.50   
...           ...              ...        ...       ...        ...   
21598  8956200760  20141013T000000   541800.0         4       2.50   
21599  7202300110  20140915T000000   810000.0         4       3.00   
21600   249000205  20141015T000000  1540000.0         5       3.75   
21605  3448900210  20141014T000000   610685.0         4       2.50   
21606  7936000429  20150326T000000  1010000.0         4       3.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
5             5420    101930     1.0           0     0  ...     11   
29            2570      7173     2.0           0     0  ...      8   
40            2570      5520     2.0           0     0  ...      9   
42            3595      5639     2.0           0     0  ...      9   
45            3160     13603     2.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21598         3118      7866     2.0           0     2  ...      9   
21599         3990      7838     2.0           0     0  ...      9   
21600         4470      8088     2.0           0     0  ...     11   
21605         2520      6023     2.0           0     0  ...      9   
21606         3510      7200     2.0           0     0  ...      9   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
5            3890           1530      2001             0    98053  47.6561   
29           2570              0      2005             0    98052  47.7073   
40           2570              0      2000             0    98074  47.6145   
42           3595              0      2014             0    98053  47.6848   
45           3160              0      2003             0    98019  47.7443   
...           ...            ...       ...           ...      ...      ...   
21598        3118              0      2014             0    98001  47.2931   
21599        3990              0      2003             0    98053  47.6857   
21600        4470              0      2008             0    98004  47.6321   
21605        2520              0      2014             0    98056  47.5137   
21606        2600            910      2009             0    98136  47.5537   

          long  sqft_living15  sqft_lot15  
5     -122.005           4760      101930  
29    -122.110           2630        6026  
40    -122.027           2470        5669  
42    -122.016           3625        5639  
45    -121.977           3050        9232  
...        ...            ...         ...  
21598 -122.264           2673        6500  
21599 -122.046           3370        6814  
21600 -122.200           2780        8964  
21605 -122.167           2520        6023  
21606 -122.398           2050        6200  

[2157 rows x 21 columns]

Now we will look for the houses with a sqft_lot area larger that li_lot: we will store it into a dataframe called df_lo:

In [122]:
df_lo=df_2000.loc[df_2000['sqft_lot']>lo_mean].copy()

In [121]:
df_lo

id             date      price  bedrooms  bathrooms  \
5      7237550310  20140512T000000  1230000.0         4        4.5   
45     8035350320  20140718T000000   488000.0         3        2.5   
79     1531000030  20150323T000000   720000.0         4        2.5   
114    8961960160  20141028T000000   480000.0         4        2.5   
147    3224079105  20140806T000000   430000.0         2        2.5   
...           ...              ...        ...       ...        ...   
21532   324069112  20140617T000000  1330000.0         4        4.0   
21540   715010530  20150113T000000  1880000.0         5        3.5   
21548  8835770330  20140819T000000  1060000.0         2        1.5   
21551  1561750040  20141224T000000  1380000.0         5        4.5   
21553  2522059251  20150409T000000   465000.0         3        2.5   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
5             5420    101930     1.0           0     0  ...     11   
45            3160     13603     2.0           0     0  ...      8   
79            3450     39683     2.0           0     0  ...     10   
114           3230     16171     2.0           0     3  ...      9   
147           2420     60984     2.0           0     0  ...      7   
...            ...       ...     ...         ...   ...  ...    ...   
21532         4420     16526     2.0           0     0  ...     11   
21540         4410     13000     2.0           0     3  ...     10   
21548         2370    184231     2.0           0     0  ...     11   
21551         4350     13405     2.0           0     0  ...     11   
21553         2050     15035     2.0           0     0  ...      9   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
5            3890           1530      2001             0    98053  47.6561   
45           3160              0      2003             0    98019  47.7443   
79           3450              0      2002             0    98010  47.3420   
114          2520            710      2001             0    98001  47.3183   
147          2420              0      2007             0    98027  47.5262   
...           ...            ...       ...           ...      ...      ...   
21532        4420              0      2013             0    98075  47.5914   
21540        2920           1490      2014             0    98006  47.5382   
21548        2370              0      2005             0    98045  47.4543   
21551        4350              0      2014             0    98074  47.6018   
21553        2050              0      2006             0    98042  47.3619   

          long  sqft_living15  sqft_lot15  
5     -122.005           4760      101930  
45    -121.977           3050        9232  
79    -122.025           3350       39750  
114   -122.253           2640        8517  
147   -121.943           1940      193842  
...        ...            ...         ...  
21532 -122.027           3510       50447  
21540 -122.111           5790       12969  
21548 -121.778           3860      151081  
21551 -122.060           3990        7208  
21553 -122.122           1300       15836  

[460 rows x 21 columns]

In [120]:
df_lo['price'].mean()

1077252.8782608695

In [119]:
df_lo['price'].median()

836500.0

In [118]:
df_li['price'].mean()

841047.5187760779

In [117]:
df_li['price'].median()

725000.0

In [116]:
df_exp[list_columns]

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0   7397300170  3710000.0         4       3.50         5550     28078     2.0   
1   9208900037  6890000.0         6       7.75         9890     31374     2.0   
2   3225069065  3080000.0         4       5.00         4550     18641     1.0   
3   6065300370  4210000.0         5       6.00         7440     21540     2.0   
4   7738500731  4500000.0         5       5.50         6640     40014     2.0   
5   3625049042  3640000.0         5       6.00         5490     19897     2.0   
6   3025059093  3100000.0         5       5.25         5090     23669     2.0   
7   4389201095  3650000.0         5       3.75         5020      8694     2.0   
8   6447300265  4000000.0         4       5.50         7080     16573     2.0   
9   8907500070  5350000.0         5       5.00         8000     23985     2.0   
10  1176001310  2950000.0         5       4.50         4340      5722     3.0   
11  5316100780  2580000.0         4       3.50         3280      3800     2.0   
12  1732800310  2000000.0         4       3.75         2870      4500     2.0   
13  8964800330  3000000.0         4       3.75         5090     14823     1.0   
14  2525049148  3420000.0         5       5.00         5450     20412     2.0   
15  2424059174  2000000.0         4       3.25         5640     35006     2.0   

    condition  grade  yr_built  waterfront  view  
0           4     12      2000           0     2  
1           3     13      2001           0     4  
2           3     10      2002           1     4  
3           3     12      2003           0     0  
4           3     12      2004           1     4  
5           3     12      2005           0     0  
6           3     12      2006           0     0  
7           3     12      2007           0     1  
8           3     12      2008           0     0  
9           3     12      2009           0     4  
10          3     10      2010           0     4  
11          3     11      2011           0     0  
12          3     10      2012           0     3  
13          3     11      2013           0     0  
14          3     11      2014           0     0  
15          3     11      2015           0     2

In [135]:
df_maxbd[list_columns]

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0   4139500200  1380000.0         6       4.50         5740     10312     2.0   
1    424069271   936000.0         7       3.75         5100     21802     2.0   
2   3425059222  1300000.0         6       3.50         6563     32670     2.0   
3   7202331220   635000.0         6       2.50         3880      5700     2.0   
4   1778360150  1240000.0         7       5.50         6630     13782     2.0   
5   7852011070  1140000.0         6       3.75         5960     20197     2.0   
6   3021059304   300000.0         6       3.00         2744      9926     2.0   
7   3185600055   495000.0         6       5.00         3440      4500     2.0   
8    627300145  1150000.0        10       5.25         4590     10920     1.0   
9   2149800278   343000.0         6       5.00         2732      7655     2.0   
10  1853081000   820000.0         5       2.75         2830      6137     2.0   
11  6891100260   830000.0         5       3.50         3700      5400     2.0   
12  5609000311   729999.0         6       4.50         3600      6110     2.0   
13  4006000183   450000.0         7       4.00         3150      7800     2.0   
14  2402100575  1130000.0         6       3.75         3010      4360     2.0   
15  9520900210   614285.0         5       2.75         2730      6401     2.0   

    condition  grade  yr_built  waterfront  view     long      lat  
0           3     11      2000           0     2 -122.110  47.5533  
1           3     10      2001           0     0 -122.040  47.5950  
2           3     10      2002           0     0 -122.157  47.6078  
3           3      7      2003           0     0 -122.038  47.6816  
4           3     10      2004           0     0 -122.118  47.5399  
5           3     10      2005           0     4 -121.869  47.5398  
6           3      7      2006           0     0 -122.216  47.2773  
7           3      8      2007           0     0 -122.219  47.4871  
8           3      9      2008           0     2 -122.113  47.5861  
9           3      7      2009           0     0 -122.211  47.3045  
10          3      9      2010           0     0 -122.058  47.5932  
11          3      9      2011           0     0 -122.117  47.7085  
12          3      9      2012           0     0 -122.291  47.5687  
13          3      8      2013           0     0 -122.279  47.5259  
14          3      9      2014           0     0 -122.333  47.6873  
15          3      8      2015           0     0 -122.160  47.7685

Looking at the results we can lift the following conclusions:
- The number of bedrooms, bathrooms and floors does not have a direct relation to the prices of the house since the gap between the most expensive houses and the others is high
- The living area size has a higher influence on the prices thant the lot area size
- The grade has influence on the prices: whenever there is 2 houses with close area size the grade does make a difference on the price


## Part 2

This part consist of defining a function that calculate the distance between 2 geographical points: It taskes the longitude and latitude as inputs:

In [129]:
from math import sin, cos, sqrt, atan2
def distance(lat1,lon1,lat2,lon2):
    R = 6373.0
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = (sin(dlat/2))**2 + cos(lat1) * cos(lat2) * (sin(dlon/2))**2
    c = 2 * atan2(sqrt(a), sqrt(1-a))
    distance = R * c
    return distance

Next we will need to extract the long and lat given an ID:

In [138]:
## df_2000.loc[(df['id']==7397300170),'long'].values[0]
def distance_id(ID1,ID2,df):
    lat1=df.loc[(df['id']==ID1),'lat'].values[0]
    long1=df.loc[(df['id']==ID1),'long'].values[0]
    lat2=df.loc[(df['id']==ID2),'lat'].values[0]
    long2=df.loc[(df['id']==ID2),'long'].values[0]
    return distance(lat1,long1,lat2,long2)
    


    

Now we will take a case where we choose 2 houses from the same year (ex: 2013) and try to calculate their distances and conclude the effect on the price:

In [145]:
df_2000[df_2000['yr_built']==2013].head(20)[list_columns]

id      price  bedrooms  bathrooms  sqft_living  sqft_lot  \
320     325059286   819900.0         5       2.75         3150      7119   
363     925059078   604950.0         3       2.50         2110      5608   
902    2296500136   839900.0         4       3.50         3810     13592   
1224   3331000070   735000.0         4       2.50         2820      6180   
1532   8165500830   409900.0         3       2.50         1690      1200   
1931   5379804150   598800.0         6       4.00         4470     17877   
2790    824059331  1610000.0         5       3.75         3530     13260   
3123   1023059511   527000.0         4       2.50         2900      6736   
3748   4389201250  2450000.0         5       4.00         4430      9000   
5025   3362401611  1170000.0         4       3.75         3920      4500   
5643   1041440050   359950.0         5       2.75         2844      3990   
6238     40000235   380000.0         5       2.50         2130      8428   
6996   3158500460   359500.0         3       2.50         2070      4689   
8120   4006000183   450000.0         7       4.00         3150      7800   
8390   2767602094   565000.0         3       2.25         1520      1221   
9531   1137450120   487500.0         4       2.50         2810      6296   
9647   5634500234   554990.0         3       2.50         2100      6092   
9986   1723049624   330000.0         5       3.00         2100      7715   
11025  2625059301   760000.0         4       3.25         2590      3889   
12509   249000180  1890000.0         4       4.25         4285      9345   

       floors  condition  grade  yr_built  waterfront  view     long      lat  
320       2.0          3      9      2013           0     0 -122.151  47.6759  
363       1.0          3      8      2013           0     0 -122.184  47.6743  
902       1.0          3      9      2013           0     1 -122.200  47.5134  
1224      2.0          3      9      2013           0     0 -122.281  47.5529  
1532      2.0          3      8      2013           0     0 -122.367  47.5389  
1931      3.0          3      9      2013           0     3 -122.273  47.4514  
2790      2.0          3     10      2013           0     0 -122.205  47.5761  
3123      2.0          3      8      2013           0     0 -122.152  47.4954  
3748      2.0          3     10      2013           0     0 -122.216  47.6168  
5025      3.0          3      8      2013           0     0 -122.346  47.6805  
5643      2.0          3      8      2013           0     0 -122.167  47.3259  
6238      2.0          3      7      2013           0     0 -122.282  47.4726  
6996      2.0          3      8      2013           0     0 -122.056  47.3545  
8120      2.0          3      8      2013           0     0 -122.279  47.5259  
8390      3.0          3      8      2013           0     0 -122.377  47.6740  
9531      2.0          3      9      2013           0     0 -122.151  47.5019  
9647      2.0          3      8      2013           0     0 -122.239  47.7508  
9986      1.0          3      7      2013           0     0 -122.319  47.4866  
11025     3.0          3      9      2013           0     0 -122.142  47.6259  
12509     2.0          3     10      2013           0     0 -122.199  47.6332

We will choose an example of 2 houses where the price range is high although the area size and the other parameters are quite close. Ex:
ID= 5379804150 Price=598800 and ID=325059286 Price=819900:


In [146]:
distance_id(5379804150,325059286,df_2000)


1593.2315387348026

The distance is 1593 km. It looks like the 2 houses are on different areas where one is more expensive than the other.

If we take another 2 houses where the prices are respectively 3110000 and 3165000:

In [150]:
distance_id(3025059093,4389201095,df_2000)

97.68673418999757

The distance is 97 km which is relatively close. We can conclude that the prices in this area are considered high

## Part 3

The at is used to access a single value from a dataframe, using the row index and the column name:

In [153]:
df_exp.at[1,'id'] # to access the id of the second row (0 is first)

9208900037

The iat is used to access a single value from a dataframe, using the row index and the column index:

In [155]:
df_exp.iat[1,0] # to access the value at the second row and first column

9208900037

## Part 4

Convert the date column from df to a datatime:

In [156]:
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

In [158]:
df['date'] = pd.to_datetime(df['date'])

In [159]:
df

id       date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520 2014-10-13  221900.0         3       1.00         1180   
1      6414100192 2014-12-09  538000.0         3       2.25         2570   
2      5631500400 2015-02-25  180000.0         2       1.00          770   
3      2487200875 2014-12-09  604000.0         4       3.00         1960   
4      1954400510 2015-02-18  510000.0         3       2.00         1680   
...           ...        ...       ...       ...        ...          ...   
21608   263000018 2014-05-21  360000.0         3       2.50         1530   
21609  6600060120 2015-02-23  400000.0         4       2.50         2310   
21610  1523300141 2014-06-23  402101.0         2       0.75         1020   
21611   291310100 2015-01-16  400000.0         3       2.50         1600   
21612  1523300157 2014-10-15  325000.0         2       0.75         1020   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
0          5650     1.0           0     0  ...      7        1180   
1          7242     2.0           0     0  ...      7        2170   
2         10000     1.0           0     0  ...      6         770   
3          5000     1.0           0     0  ...      7        1050   
4          8080     1.0           0     0  ...      8        1680   
...         ...     ...         ...   ...  ...    ...         ...   
21608      1131     3.0           0     0  ...      8        1530   
21609      5813     2.0           0     0  ...      8        2310   
21610      1350     2.0           0     0  ...      7        1020   
21611      2388     2.0           0     0  ...      8        1600   
21612      1076     2.0           0     0  ...      7        1020   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0                  0      1955             0    98178  47.5112 -122.257   
1                400      1951          1991    98125  47.7210 -122.319   
2                  0      1933             0    98028  47.7379 -122.233   
3                910      1965             0    98136  47.5208 -122.393   
4                  0      1987             0    98074  47.6168 -122.045   
...              ...       ...           ...      ...      ...      ...   
21608              0      2009             0    98103  47.6993 -122.346   
21609              0      2014             0    98146  47.5107 -122.362   
21610              0      2009             0    98144  47.5944 -122.299   
21611              0      2004             0    98027  47.5345 -122.069   
21612              0      2008             0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21608           1530        1509  
21609           1830        7200  
21610           1020        2007  
21611           1410        1287  
21612           1020        1357  

[21613 rows x 21 columns]